CMakeLists.txt

```diff
@@ -95,162 +142,154 @@ macro(clangformat_targets)
-  endif()
+  set(CMAKE_REQUIRED_LINK_OPTIONS -fuse-ld=lld)
   check_cxx_source_compiles("int main() { return 1; }" LLD_WORKS)
+  if (LLD_WORKS)
+    message(STATUS "Using LLD to link snmalloc shims")
+  endif()
 
-  macro(add_shim name type)
+  function(add_shim name type)
     add_library(${name} ${type} ${ARGN})
-    target_link_libraries(${name} snmalloc_lib)
-    if(NOT MSVC)
-      target_compile_definitions(${name} PRIVATE "SNMALLOC_EXPORT=__attribute__((visibility(\"default\")))")
-    endif()
+    target_link_libraries(${name} snmalloc)
     set_target_properties(${name} PROPERTIES CXX_VISIBILITY_PRESET hidden)
+    target_compile_definitions(${name} PRIVATE "SNMALLOC_USE_${SNMALLOC_CLEANUP}")
 
-    # Ensure that we do not link against C++ stdlib when compiling shims.
-    if(NOT MSVC)
-      set_target_properties(${name} PROPERTIES LINKER_LANGUAGE C)
-      if (LLD_WORKS)
-        message("Using LLD.")
+    if(MSVC)
+      target_compile_options(${name} PRIVATE /Zi /W4 /WX /wd4127 /wd4324 /wd4201)
+      set(CMAKE_EXE_LINKER_FLAGS_RELEASE "${CMAKE_EXE_LINKER_FLAGS_RELEASE} /DEBUG")
+    else()
+      target_compile_definitions(${name} PRIVATE "SNMALLOC_EXPORT=__attribute__((visibility(\"default\")))")
+      target_compile_options(${name} PRIVATE
+        -fno-exceptions -fno-rtti -fomit-frame-pointer -ffunction-sections
+        -Wall -Wextra -Werror -Wundef
+        $<$<CXX_COMPILER_ID:Clang>:-Wsign-conversion -Wconversion>)
+      # Static TLS model is unsupported on Haiku.
+      # All symbols are always dynamic on haiku and -rdynamic is redundant (and unsupported).
+      if (NOT CMAKE_SYSTEM_NAME STREQUAL "Haiku")
+        target_compile_options(${name} PRIVATE -ftls-model=initial-exec)
+        target_compile_options(${name} PRIVATE $<$<BOOL:${SNMALLOC_CI_BUILD}>:-g>)
+      endif()
 
-        # Only include the dependencies that actually are touched.
-        # Required so C++ library is not included with direct pthread access.
-        target_link_libraries(${name} -Wl,--as-needed)
+      if(SNMALLOC_OPTIMISE_FOR_CURRENT_MACHINE)
+        check_cxx_compiler_flag(-march=native SUPPORT_MARCH_NATIVE)
+        if (SUPPORT_MARCH_NATIVE)
+          target_compile_options(${name} -march=native)
+        else()
+          message(WARNING "Compiler does not support `-march=native` required by SNMALLOC_OPTIMISE_FOR_CURRENT_MACHINE")
+        endif()
+      endif()
 
-        # Remove all the duplicate new/malloc and free/delete definitions
-        target_link_libraries(${name} -Wl,--icf=all -fuse-ld=lld)
+      # Ensure that we do not link against C++ stdlib when compiling shims.
+      # If the compiler supports excluding the C++ stdlib implementation, use
+      # it.  Otherwise, fall back to linking the library as if it were C, which
+      # has roughly the same effect.
+      if (NOT ${SNMALLOC_CLEANUP} STREQUAL CXX11_DESTRUCTORS)
+        check_linker_flag(CXX "-nostdlib++" SNMALLOC_LINKER_SUPPORT_NOSTDLIBXX)
```

Comment:
I guess this is better than `--as-needed` as we will get link failures if we accidentally include something that uses C++.  But this might make Debugging harder, as adding anything like `std::cout` will then fail to compile.   Could we still include the `--as-needed` in the else branch.

Reply:
Note that this isn't applied to the tests: they still link the C++ standard library and so can use whatever they want.  This just prevents the shim libraries from depending on lib[std]c++.  I've hit problems in the past where we accidentally used something like `std::cout` on a code path that ended up being called before malloc was initialised and so we got deadlock rather than useful crashes so I'd much prefer we got a linker failure from this rather than a difficult-to-debug run-time condition.

Reply:
I agree entirely, that having a hard error for including the stdlib++ most of the time is the right default.  Prior to this change I was thinking about something like 
```sh
objdump -p libsnmallocshim.so | grep NEEDED
```

Perhaps, we could have command line option to enable stdlib++ (i.e. disable this code). I am just thinking about times where I have been trying to debug `LD_PRELOAD`ed things, and I carefully added some bits of logging.

I'm also concerned if we add stats printing on teardown back as an option then stdlib++ might be needed.

Reply:
I've added the linker flag that checks for undefined symbols and errors at link time with something more useful.  If you're doing some local tests with these things then you can patch this out.  We can also add the runtime dependency back in explicitly if and only if we enable stats so that we'll still get CI failures if we use anything from the C stdlib in non-stats builds.